// @sentry/integrations/cjs/index.js
Object.defineProperty(exports, '__esModule', { value: true });

const captureconsole = require('./captureconsole.js');
const debug = require('./debug.js');
const dedupe = require('./dedupe.js');
const extraerrordata = require('./extraerrordata.js');
const offline = require('./offline.js');
const reportingobserver = require('./reportingobserver.js');
const rewriteframes = require('./rewriteframes.js');
const sessiontiming = require('./sessiontiming.js');
const transaction = require('./transaction.js');
const httpclient = require('./httpclient.js');
const contextlines = require('./contextlines.js');



exports.CaptureConsole = captureconsole.CaptureConsole;
exports.captureConsoleIntegration = captureconsole.captureConsoleIntegration;
exports.Debug = debug.Debug;
exports.debugIntegration = debug.debugIntegration;
exports.Dedupe = dedupe.Dedupe;
exports.dedupeIntegration = dedupe.dedupeIntegration;
exports.ExtraErrorData = extraerrordata.ExtraErrorData;
exports.extraErrorDataIntegration = extraerrordata.extraErrorDataIntegration;
exports.Offline = offline.Offline;
exports.ReportingObserver = reportingobserver.ReportingObserver;
exports.reportingObserverIntegration = reportingobserver.reportingObserverIntegration;
exports.RewriteFrames = rewriteframes.RewriteFrames;
exports.rewriteFramesIntegration = rewriteframes.rewriteFramesIntegration;
exports.SessionTiming = sessiontiming.SessionTiming;
exports.sessionTimingIntegration = sessiontiming.sessionTimingIntegration;
exports.Transaction = transaction.Transaction;
exports.HttpClient = httpclient.HttpClient;
exports.httpClientIntegration = httpclient.httpClientIntegration;
exports.ContextLines = contextlines.ContextLines;
exports.contextLinesIntegration = contextlines.contextLinesIntegration;
